docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,259 @@
+  ObjectMeta           `json:"metadata,omitempty"`
+
+  // ProvisionerType indicates the type of the provisioner.
+  ProvisionerType string `json:"provisionerType,omitempty"`
+
+  // Parameters for dynamic volume provisioner.
+  ProvisionerParameters map[string]string `json:"provisionerParameters,omitempty"`
+}
+
+```
+
+`PersistentVolumeClaimSpec` and `PersistentVolumeSpec` both get Class attribute
+(the existing annotation is used during incubation):
+
+```go
+type PersistentVolumeClaimSpec struct {
+    // Name of requested storage class. If non-empty, only PVs with this
+    // pv.Spec.Class will be considered for binding and if no such PV is
+    // available, StorageClass with this name will be used to dynamically
+    // provision the volume.
+    Class string
+...
+}
+
+type PersistentVolumeSpec struct {
+    // Name of StorageClass instance that this volume belongs to.
+    Class string
+...
+}
+```
+
+Storage classes are natural to think of as a global resource, since they:
+
+1.  Align with PersistentVolumes, which are a global resource
+2.  Are administrator controlled
+
+### Provisioning configuration
+
+With the scheme outlined above the provisioner creates PVs using parameters specified in the `StorageClass` object. 
+
+### Provisioner interface changes
+
+`struct volume.VolumeOptions` (containing parameters for a provisioner plugin)
+will be extended to contain StorageClass.ProvisionerParameters.
+
+The existing provisioner implementations will be modified to accept the StorageClass configuration object.
+
+### PV Controller Changes
+
+The persistent volume controller will be modified to implement the new
+workflow described in this proposal.  The changes will be limited to the
+`provisionClaimOperation` method, which is responsible for invoking the
+provisioner and to favor existing volumes before provisioning a new one.
+
+## Examples
+
+### AWS provisioners with distinct QoS
+
+This example shows two storage classes, "aws-fast" and "aws-slow".
+
+```
+apiVersion: v1
+kind: StorageClass
+metadata:
+  name: aws-fast
+provisionerType: kubernetes.io/aws-ebs
+provisionerParameters:
+   zone: us-east-1b
+   type: ssd
+
+
+apiVersion: v1
+kind: StorageClass
+metadata:
+  name: aws-slow
+provisionerType: kubernetes.io/aws-ebs
+provisionerParameters:
+   zone: us-east-1b
+   type: spinning
+```
+
+# Additional Implementation Details
+
+0. Annotation `volume.alpha.kubernetes.io/storage-class` is used instead of `claim.Spec.Class` and `volume.Spec.Class` during incubation.
+
+1. `claim.Spec.Selector` and `claim.Spec.Class` are mutually exclusive. User can either match existing volumes with `Selector` XOR match existing volumes with `Class` and get dynamic provisioning by using `Class`. This simplifies initial PR and also provisioners. 
+
+# Cloud Providers
+Since the `volume.alpha.kubernetes.io/storage-class` is in use a `StorageClass` must be defined to support provisioning.  No default is assumed as before.
+
```

Comment:
We should add a "future work" section:

- default provisioner
- selectors and classes together
- self-describing parameters to provisioners
- out-of-tree provisioners